### PR TITLE
usb: device_next: add few more string descriptor helpers

### PR DESCRIPTION
**Pull request by jfischer-no**
usb: device_next: add few more string descriptor helpers

Add function to get string descriptor index and function
to remove linked descriptor from a device. This abstracts
it a bit so that the user does not need to know how it is
handled internally.

usb: device_next: add string descriptor for CDC ACM

Add a string descriptor whose index is used for IAD and interface
descriptors. The string is derived from the virtual CDC ACM UART
nodelabel, which allows the user to accurately identify the UART
interface on the device side.
It can probably be further improved by converting the string to
upper case and replacing the underline with a space.

### CODE REVIEW

File: subsys/usb/device_next/class/usbd_cdc_acm.c [hunk 1]
```diff
@@ -394,14 +396,36 @@ static int usbd_cdc_acm_ctd(struct usbd_class_node *const c_nd,
...
 	desc->iad_cdc.bFirstInterface = desc->if0.bInterfaceNumber;
 	desc->if0_union.bControlInterface = desc->if0.bInterfaceNumber;
 	desc->if0_union.bSubordinateInterface0 = desc->if1.bInterfaceNumber;
 
+	if (usbd_add_descriptor(c_nd->data->uds_ctx, data->iad_desc_nd)) {
+		LOG_ERR("Failed to add iFunction string descriptor");
+	} else {
+		desc->iad_cdc.iFunction = usbd_get_descriptor_idx(data->iad_desc_nd);
+		desc->if0.iInterface = usbd_get_descriptor_idx(data->iad_desc_nd);
+		desc->if1.iInterface = usbd_get_descriptor_idx(data->iad_desc_nd);
```
Comment (tmon-nordic) [inline subsys/usb/device_next/class/usbd_cdc_acm.c:411]: I don't quite like using the same string for all three. I think only the iFunction should be set here.

File: subsys/usb/device_next/class/usbd_cdc_acm.c [hunk 2]
```diff
@@ -1102,6 +1127,9 @@ static struct usbd_cdc_acm_desc cdc_acm_desc_##n = {				\
 		     " is not assigned to a USB device controller");		\
 										\
 	CDC_ACM_DEFINE_DESCRIPTOR(n);						\
+	USBD_DESC_STRING_DEFINE(iad_desc_nd_##n,				\
+				DT_NODE_FULL_NAME(DT_DRV_INST(n)),		\
```
Comment (tmon-nordic) [inline subsys/usb/device_next/class/usbd_cdc_acm.c:1131]: I don't really like using node name as a string descriptor. The identifiers used in code are frequently not what the end users should see.

USB provides multiple language string descriptor support. The fact that in practice only a single locale tends to be used in many devices is separate concern. However, we should really have some framework for managing string descriptors. Deriving string descriptor from node label is pretty bad solution in my opinion.

**Review by tmon-nordic — CHANGES_REQUESTED** [review]

File: subsys/usb/device_next/class/usbd_cdc_acm.c (hunk 1 above)
Comment (jfischer-no) [reply subsys/usb/device_next/class/usbd_cdc_acm.c:411]: That was my first thought too, but I am afraid there are users who expect it (also) at interface level.

File: subsys/usb/device_next/class/usbd_cdc_acm.c (hunk 2 above)
Comment (tmon-nordic) [reply subsys/usb/device_next/class/usbd_cdc_acm.c:1131]: Being only way to accurately identify the interface does not make it good for string descriptor.

File: subsys/usb/device_next/class/usbd_cdc_acm.c (hunk 1 above)
Comment (tmon-nordic) [reply subsys/usb/device_next/class/usbd_cdc_acm.c:411]: I think the users would expect to be able to set some meaningful interface string descriptors. Having the same string descriptor assigned to both interfaces and function is not really making things better IMHO.

**Comment by jfischer-no** [comment]
Outdated

**Pull request closed without merging**